src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/NodeFactory.cs

```diff
@@ -869,19 +869,26 @@ public TypeDesc ArrayOfTEnumeratorType
         }
 
         private TypeDesc _systemICastableType;
+        private bool? _hasSystemICastableType;
```

Comment:
I think it would be more elegant to delete the existing `ICastableInterface` property and replacing it with:

```csharp
private object _systemICastableTypeOrPlaceholder;

public bool IsICastableInterface(TypeDesc type)
{
    if (_systemICastableTypeOrPlaceholder == null)
    {
        _systemICastableTypeOrPlaceholder = _context.SystemModule.GetType("System.Runtime.CompilerServices", "ICastable", false) ?? new object();
    }
    return type == _systemICastableTypeOrPlaceholder;
}
```

Then replace all uses of the deleted property with this new code.